### ISSUE TITLE
Redact the Initial Version of the Documentation

### ISSUE THREAD

**Issue by VincenzoFerme** [benchflow/benchflow#51]
TODO: Organise the following notes to a better structure.

- working with branches with pull request to the devel branch that is locked
- naming for branches and granularity of commits (add the feature-label-refactoring- branch you were proposing, I agree it makes sense. Under the feature-label- branch section)
- when to open an issues and when to place a TODO in the code
- labels for issues (and color scheme + naming scheme of labels), versioning for the different projects
- what has to be ignored when committing:
  - ide specific artifacts
  - compiled code (e.g. .class)
- how to share or not share secrets (e.g. passwords)
- how to import the projects in the IDE and the fact they all have to be maven projects (if they are scala and/or Java projects) and guidelines for a clean Pom definition
- what is the standard structure for library and for services in Java/scala
- how to define tests, which tests and how to work with test driven development (just enough tests but at all layers) 
- how to document APIs in the code and on the doc
- how to bump a version for a project and for benchflow and when to do so
- how the continuous integration and delivery works ??
- Structure of development/BenchFlow environment
- Exceptions
- Architecture (with different levels)

@jeppe-style did I forget something you think is useful to correctly work on the repo? The things on the continuous integration I'll try to have them ready over the weekend. I will also define places for different notes and discussion in a more formal way, as well as  I will propose a structure for the documentation.  Everything that is not clear we should clarify and document so it is there and it is something we agree on. Once I get you feedback, I'll open an issue about this, so we can reference the ongoing work.

@VincenzoFerme cross check with ASQ devops

**Comment by igormoreno** [comment]
It seems that some of these the items were already completed (looking at https://github.com/VincenzoFerme/benchflow-temporary-documentation/blob/master/development/contributing-to-benchflow.md). Which ones? @jeppe-style  Can you check that? We should have a better way of tracking these things. Maybe break down some of these items into other issues?

**Comment by igormoreno** [comment]
if we are moving everything to a single repository. Shouldn't we have also the documentation there?

**Comment by jeppe-style** [comment]
@igormoreno yes, Vincenzo has setup a way that we can move the documentation to the single repo under the [docs](https://github.com/benchflow/benchflow/tree/devel/docs) folder.

Once we move things from the temporary documentation I think we can update the information here or create new smaller issues that we complete.

**Comment by VincenzoFerme** [comment]
Ongoing work on README and Docs:

## READMEs

- [ ] BenchFlow main readme: continue TODOs, and iterate on what is there. Link also keywork to the documentation, once defined. When the project focus changes, update the main description as well., make it clear the use of Issues also for development and not only for bugs, mention where we do release artefacts etc, maybe directly pointing to the docs, and just having a short summary in the README.
- [ ] Single project readme: 

## Docs

- [ ] Improve and update the README
- [ ] Continue to build the main page, and the website/documentation organisation:
    - [ ] Link to social and GitHub page
    - [ ] Define the page headers in `_include\page-headers.html`
- [ ] Work on the generation so that it is correclty generated and ready for release and serving it on Github Pages, also removing artefacts that are not needed for deploy
- [ ] Add Google analytics in the header and other relevant actions, setting the onclick events in the navigation menu to tack the navigation. Add also https://github.com/igrigorik/ga-beacon to the different README files.
- [ ] Fill in with contents
    - [ ] Getting Started:
        - [ ] Introduction: present BenchFlow, state the current focus of the framework (as use cases), ...
    - [ ] User Guide:
        - [ ] [@jeppe-style](https://github.com/jeppe-style), always starting from an overview page linking all the relevant content
    - [ ] Developer Guide:
        - [ ] Information on labeling, issue and pull request (e.g., delete branch after merge if feature branch, ...) labeling, assignment and management.
        - [ ] [@jeppe-style](https://github.com/jeppe-style), always starting from an overview page linking all the relevant content
- [ ] Layout:
    - [ ] Vertical following navigation bar on the left
    - [ ] Got to top at the bottom
    - [ ] Mobile: nicer approach to show the menu, maybe with a dropdown
- [ ] Check changes of the Bootstrap example, and schedule regular checks in case it is needed
- [ ] milestone selection to the doc and name changing
- [ ] Clean up not needed js (also from the `package.json` in the root), css and other layout/functionality files

**Comment by VincenzoFerme** [comment]
@igormoreno @jeppe-style I just committed a first draft for the documentation structure. You can see it in: https://github.com/benchflow/benchflow/tree/feature-docs-draft-structure. It also contains some notes (as comment in the markdown files) about the content I would like to add in the different part of the documentation. 

I'd like to hear your feedback on the structure, and plan the work to update it with content. I took into account all the material we have in the temporary documentation as well, while figuring out the structure of the draft.

**Comment by jeppe-style** [comment]
To me it looks quite complete at a first glance. I think it would be good to have (not only as a comment) a description of each section (at least a sentence). It would be easier to navigate like that and also for us to give feedback.

**User Guide:** I think a section about the DSL and all its properties (and examples of these properties) would be good to have. I think now it is split up into several sections - it is not so clear.

**Developer Guide:** Maybe the sections about contributing could be put together. Now I see there are some parts that given an overview of the system and then some other sections more about how to actually contribute in the right way.

**Comment by VincenzoFerme** [comment]
Thanks @jeppe-style. For what concern having a description, I agree and we can do it section by section while starting to add the first draft of the content. 

**User Guide**: The idea is that in `Anatomy of a Test` we describe what it is possible to do and we provide links to the different self-contained sections we have, then in `Simple Test` we provide examples on how to execute a single test while in `Objective Driven Test` we provide examples about more advanced testing. Finally in `Complete Test Examples` we have complete examples covering everything discussed in the dedicated sections. The reason behind having multiple sections is that the DSL is complex, and having all in one place might overload the user.

**Developer Guide**: This is wanted, because not all the reader of the developer version might want to contribute. The the sections about the overview of the system, we can point also to a specific subsection of the sections related to how contribute to the codebase, if needed.

**Comment by VincenzoFerme** [comment]
Migrate everything relevant from the documentation from: https://github.com/VincenzoFerme/benchflow-temporary-documentation, then this repo can also be closed in favour of using in progress / need discussion issues with attached documentation that can be updated, or even better documentation branches tracked in issues so that to iterate on the documentation until it is ready for merging.

**Comment by VincenzoFerme** [comment]
Something relevant for how to build:

- Before a pull request get merged we need to enforce that a complete build of all the projects that got changed in the branch get a build, so that this is reported in the pull request status. This happens (due to our settings) when devel is merged into the branch, otherwise we need to setup an automated trigger enabling us to do so before the merge (maybe from the command line utility described in: https://github.com/benchflow/benchflow/issues/325, we can run locally the same code that on wercker runs after a merge of devel into the branch).

SLACK:

Check how to verify who is following a thread, otherwise need to ping. Or enforce rule that if last message older than a given time, then mention and possibly also working offline.

**Comment by VincenzoFerme** [comment]
Complete and link to the official documentation:

- https://github.com/benchflow/benchflow/pull/429
- https://github.com/benchflow/benchflow/pull/430

Everything has to be versionable, including [diagrams](http://www.gravizo.com), and we need to provide hints and best tools for working with these tools.
